xonsh/shell.py

```diff
@@ -128,7 +115,15 @@ def default(self, line):
         if code is None:
             return
         try:
-            self.execer.exec(code, mode='single', glbs=self.ctx)  # no locals
+            # Temporarily redirect stdout and stderr to save results in
+            # history.
+            with redirect_stdout(self.stdout):
+                with redirect_stderr(self.stderr):
```

Comment:
I think that you can actually write 

```python
with redirect_stdout(self.stdout), redirect_stderr(self.stderr):
```

and save a level of nesting, for what it is worth.